src/HotChocolate/Core/src/Types.Scalars/PostalCodeType.cs

```diff
@@ -0,0 +1,156 @@
+using System;
+using System.Linq;
+using System.Text.RegularExpressions;
+using HotChocolate.Language;
+
+namespace HotChocolate.Types.Scalars
+{
+    /// <summary>
+    /// The PostalCode scalar type represents a valid postal code.
+    /// </summary>
+    public class PostalCodeType : StringType
+    {
+        /// <summary>
+        /// US - United States
+        /// UK - United Kingdom
+        /// DE - Germany
+        /// CA - Canada
+        /// FR - France
+        /// IT - Italy
+        /// AU - Australia
+        /// NL - Netherlands
+        /// ES - Spain
+        /// DK - Denmark
+        /// SE - Sweden
+        /// BE - Belgium
+        /// IN - India
+        /// AT - Austria
+        /// PT - Portugal
+        /// CH - Switzerland
+        /// LU - Luxembourg
+        /// </summary>
+        private static readonly string[] _validationPatterns =
+        {
+            ScalarResources.PostalCode_ValidationPattern_US,
+            ScalarResources.PostalCode_ValidationPattern_DE,
+            ScalarResources.PostalCode_ValidationPattern_UK,
+            ScalarResources.PostalCode_ValidationPattern_CA,
+            ScalarResources.PostalCode_ValidationPattern_FR,
+            ScalarResources.PostalCode_ValidationPattern_IT,
+            ScalarResources.PostalCode_ValidationPattern_AU,
+            ScalarResources.PostalCode_ValidationPattern_NL,
+            ScalarResources.PostalCode_ValidationPattern_ES,
+            ScalarResources.PostalCode_ValidationPattern_DK,
+            ScalarResources.PostalCode_ValidationPattern_SE,
+            ScalarResources.PostalCode_ValidationPattern_BE,
+            ScalarResources.PostalCode_ValidationPattern_IN,
+            ScalarResources.PostalCode_ValidationPattern_AT,
+            ScalarResources.PostalCode_ValidationPattern_PT,
+            ScalarResources.PostalCode_ValidationPattern_CH,
+            ScalarResources.PostalCode_ValidationPattern_LU
+        };
+
+        /// <summary>
+        /// Start with a limited set as suggested here:
+        /// http://www.pixelenvision.com/1708/zip-postal-code-validation-regex-php-code-for-12-countries/
+        /// and here:
+        /// https://stackoverflow.com/questions/578406/what-is-the-ultimate-postal-code-and-zip-regex
+        /// </summary>
+        private static readonly Func<string, bool> _validPostalCode = input =>
+            _validationPatterns.Select(pattern => Regex.Match(
```

Comment:
In `hotpaths` like scalars we try to avoid Linq as it allocated objects every time it is used.
We even try to avoid foreach loops because they create a enumerator
the best thing is always a for loop